File: ordenamiento/quick.py
##vamos a programar el algoritmo quicksort
#O(n log n)
import random
def quicksort(arreglo):
    if len(arreglo) <= 1:
        return arreglo
    pivote = arreglo[len(arreglo) // 2] #pivote es el elemento del medio
    
    izquierda = [x for x in arreglo if x < pivote] #elementos menores al pivote
    medio = [x for x in arreglo if x == pivote] #elementos iguales al pivote
    derecha = [x for x in arreglo if x > pivote] #elementos mayores al pivote
    return quicksort(izquierda) + medio + quicksort(derecha)

n=20
arreglo = [random.randint(0, 100) for i in range(n)]
print(quicksort(arreglo))  

File: ordenamiento/quick.cpp
#include <iostream>
#include <vector>

using namespace std;

void quicksort(vector<int> &v, int left, int right) {
    if (left >= right) return;
    
    int i = left, j = right;
    double pivot = v[(left + right) / 2];
    while (i <= j) {
        while (v[i] < pivot) i++; //me muevo en la lista izquierda hasta encontrar un elemento mayor al pivote
        while (v[j] > pivot) j--; //me muevo en la lista derecha hasta encontrar un elemento menor al pivote
        if (i <= j) {
            swap(v[i], v[j]);
            i++;
            j--;
        }
    }
    if (left < j) quicksort(v, left, j);
    if (i < right) quicksort(v, i, right);
}

int main(){

    vector<int> v = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5};
    quicksort(v, 0, v.size() - 1);
    for (int i = 0; i < v.size(); i++) {
        cout << v[i] << " ";
    }
    cout << endl;
    return 0;
}